src/kernel/capinv.c

```diff
@@ -434,8 +433,12 @@ cap_switch_thd(struct pt_regs *regs, struct thread *curr, struct thread *next,
 		return 0;
 	}
 
-	copy_gp_regs(regs, &curr->regs);
-	__userregs_set(&curr->regs, 0, __userregs_getsp(&curr->regs), __userregs_getip(&curr->regs));
+	if (!(curr->state & THD_STATE_PREEMPTED)) {
```

Comment:
My analysis: In order to modify the return path in entry.S, I needed to save the registers correctly when there is a thread switch to this thread next time. If it was pre-empted, I have to have the snapshot of all the registers, and if it wasn't, I only needed GP registers to save. This is required for making a decision of either doing a IRET or SYSEXIT. 

Also, Thread-switch (co-operative switching) has improved drastically. Before all these changes, it was 468cycles, now its around 310cycles. 

I understand that you'll need to think through the implications. Will look forward to it!